source/opt/const_folding_rules.cpp

```diff
@@ -1587,6 +1697,72 @@ BinaryScalarFoldingRule FoldFTranscendentalBinary(double (*fp)(double,
         return nullptr;
       };
 }
+
+enum Sign { Signed, Unsigned };
+
+// Returns a BinaryScalarFoldingRule that applies `op` to the scalars.
+// The `signedness` is used to determine if the operands should be interpreted
+// as signed or unsigned. If the operands are signed, the value will be sign
+// extended before the value is passed to `op`. Otherwise the values will be
+// zero extended.
+template <Sign signedness>
+BinaryScalarFoldingRule FoldBinaryIntegerOperation(uint64_t (*op)(uint64_t,
+                                                                  uint64_t)) {
+  return
+      [op](const analysis::Type* result_type, const analysis::Constant* a,
+           const analysis::Constant* b,
+           analysis::ConstantManager* const_mgr) -> const analysis::Constant* {
+        assert(result_type != nullptr && a != nullptr && b != nullptr);
+        const analysis::Integer* integer_type = a->type()->AsInteger();
+        assert(integer_type != nullptr);
+        assert(integer_type == result_type->AsInteger());
+        assert(integer_type == b->type()->AsInteger());
+
+        // In SPIR-V, the signedness of the operands is determined by the
```

Comment:
I don't want to assert on invalid spir-v unless it will cause the compiler to crash. We generally want to assume the code is valid, and not repeat all of the checks that the validator does.

I've updated to be inline with what the spir-v spec says:

https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_unsigned_versus_signed_integers:~:text=Note%20in%20both%20cases%20all%20signed%20and%20unsigned%20operations%20always%20work%20on%20unsigned%20types%2C%20and%20the%20semantics%20of%20operation%20come%20from%20the%20opcode.%20SPIR%2DV%20does%20not%20know%20which%20way%20is%20being%20used%3B%20it%20is%20set%20up%20to%20support%20both%20ways%20of%20thinking.